pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1566,9 +1610,10 @@ public int reloadSegment(@Nonnull String tableNameWithType, @Nonnull String segm
   }
 
   // Check to see if the table has been explicitly configured to NOT use messageBasedRefresh.
-  private boolean shouldSendMessage(OfflineSegmentZKMetadata segmentZKMetadata) {
-    final String rawTableName = segmentZKMetadata.getTableName();
-    TableConfig tableConfig = ZKMetadataProvider.getOfflineTableConfig(_propertyStore, rawTableName);
+  private boolean shouldSendMessage(TableConfig tableConfig) {
+    if (tableConfig == null) {
+      return true;
```

Comment:
It should return true here and my new code does not change the existing logic. It returns false iff the table config explicitly sets the custom config field. If the table config is nil, true is returned.

Reply:
True should be returned instead. My refactoring does not change the current logic. False is returned iff only if tableConfig explicitly does not allow sending message (by setting the custom config field.)

Reply:
If there is no tableConfig and we are getting a segment, then the behavior should be that we throw an exception. In the "APPEND" case, we seem to have a Preconditions.checkNotNull(tableConfig).  I am fine adding that here (one step better than what used to happen before, which is NPE).
Since it is a 'segments' operation, perhaps 404 is not the best, so I am fine with Preconditions check. Please add that in the caller to this method.

If there is no helix resource, we will get an exception anyway.

If there _is_ a helix resource (but no table config), maybe we got into an inconsistent state or are in the process of dropping a table. Either way, exception is fine (and better than trying to send messages).